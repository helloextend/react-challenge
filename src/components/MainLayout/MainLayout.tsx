import React, { Component } from 'react';
import SearchBar from './SearchBar/SearchBar';
import ResultsList from './ResultsList/ResultsList';
import FavoritesList from './FavoritesList/FavoritesList';
import { Searches } from '../../assets/custom';

type DogState = {
  results: Array<string>,
  favorites: Array<string>,
  previousSearches: Searches,
}

class MainLayout extends Component<{}, DogState> {
  constructor(props: {}) {
    super(props);
    this.state = {
      results: [''],
      favorites: [''],
      previousSearches: {
        valid: [],
        invalid: []
      }
    };
  }

  componentDidMount() {
    this.setState({
      favorites: JSON.parse(localStorage.getItem('favorites')) || []
    })
  }

  handleSearchClick = (ev) => {
    const searchTerm = ev.target.searchTerm.value
    ev.preventDefault();
    fetch(`https://dog.ceo/api/breed/${searchTerm}/images`)
      .then(response => response.json())
      .then(responseJson => {
        if (responseJson.status === 'error') {
          const newSearches = {
            valid: this.state.previousSearches.valid,
            invalid: [...this.state.previousSearches.invalid, searchTerm]
          }
          this.setState({
            previousSearches: newSearches
          })
        } else {
          const newSearches = {
            valid: [...this.state.previousSearches.valid, searchTerm],
            invalid: this.state.previousSearches.invalid
          }
          this.setState({
            results: responseJson.message,
            previousSearches: newSearches
          })
        }
      })
      .catch(console.error)
  }

  handleToggleFavorite = (url) => {
    if (this.state.favorites.indexOf(url) === -1) {
      this.setState({
        favorites: [...this.state.favorites, url]
      })
    } else {
      this.setState({
        favorites: this.state.favorites.filter(item => item !== url)
      })
    }

    const newUrls = [...this.state.favorites, url]

    localStorage.setItem('favorites', JSON.stringify(newUrls))
  }


  render() {
    return (
      <>
        <SearchBar handleSearch={this.handleSearchClick} previousSearches={this.state.previousSearches} />
        <ResultsList columnCount={3} clickHeart={this.handleToggleFavorite} imgUrls={this.state.results} favorites={this.state.favorites} />
        <FavoritesList clickHeart={this.handleToggleFavorite} favorites={this.state.favorites} />
      </>
    );
  }
}



export default MainLayout;