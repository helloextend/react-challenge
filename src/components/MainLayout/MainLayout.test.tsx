import * as React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import MainLayout from './MainLayout'
import { GlobalWithFetchMock } from "jest-fetch-mock";

const customGlobal: GlobalWithFetchMock = global as GlobalWithFetchMock;
customGlobal.fetch = require('jest-fetch-mock');
customGlobal.fetchMock = customGlobal.fetch;

const testData = {
  imgUrl: 'https://images.dog.ceo/breeds/hound-afghan/n02088094_1023.jpg',
  customEvent: {
    target: {
      searchTerm: {
        value: 'hound'
      }
    },
    preventDefault: () => { }
  }
}

describe('MainLayout component', () => {
  it('renders without crashing', () => {
    shallow(<MainLayout />)

  });

  it('adds urls to favorites', () => {
    const wrapper = shallow(<MainLayout />)
    const componentInstance = wrapper.instance();
    componentInstance.componentDidMount();
    componentInstance.componentWillMount();
    componentInstance.handleToggleFavorite(testData.imgUrl);

    expect(wrapper.state('favorites')).toHaveLength(1);
  });

  it('removes urls from favorites', () => {
    const wrapper = shallow(<MainLayout />)
    const componentInstance = wrapper.instance();
    componentInstance.componentDidMount();
    componentInstance.componentWillMount();
    componentInstance.handleToggleFavorite(testData.imgUrl);
    componentInstance.handleToggleFavorite(testData.imgUrl);

    expect(wrapper.state('favorites')).toHaveLength(0);
  });


  it("makes a fetch call when user clicks search", done => {
    const wrapper = shallow(<MainLayout />);

    setTimeout(() => {
      wrapper.update();
      const componentInstance = wrapper.instance();
      componentInstance.handleSearchClick(testData.customEvent);

      expect(global.fetch).toHaveBeenCalled();

      done();
    });
  })


  it('renders the UI as expected', () => {
    const tree = renderer
      .create(<MainLayout />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

})
