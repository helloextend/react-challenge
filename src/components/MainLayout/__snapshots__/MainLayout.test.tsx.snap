// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainLayout component renders the UI as expected 1`] = `
Array [
  <form
    className="css-79elbk"
    onSubmit={[Function]}
  >
    <input
      className="css-1ich0l"
      id="searchTerm"
      type="text"
    />
    <button
      className="css-1blg6he"
      type="submit"
    >
      Search
    </button>
  </form>,
  <div
    className="css-r0qpvp"
  >
    
  </div>,
  <h1
    className="css-ag576t"
  >
    <img
      alt="red heart icon"
      className="css-19vi65q"
      onClick={[Function]}
      src="test-file-stub"
      style={
        Object {
          "marginRight": "30px",
        }
      }
    />
    Favorites
  </h1>,
  <div
    className="css-1740rzu"
  >
    <div
      className="css-79elbk"
    >
      <img
        alt="dog"
        className="css-aynx1p"
        src="https://images.dog.ceo/breeds/hound-afghan/n02088094_1023.jpg"
      />
      <img
        alt="white heart icon"
        className="css-19vi65q"
        onClick={[Function]}
        src="test-file-stub"
        style={
          Object {
            "bottom": "34px",
            "position": "absolute",
            "right": "10px",
          }
        }
      />
    </div>
  </div>,
]
`;
